### DISCUSSION TITLE
Can you alter a transformation after it has been added?

### DISCUSSION THREAD

**Discussion by BradyAJohnston** [MDAnalysis/mdanalysis#4477] [General]
I'm not _super_ familiar with the use of [transformations](https://docs.mdanalysis.org/stable/documentation_pages/transformations/positionaveraging.html), but once you add them to a universe, can they be altered or removed from that universe?

**Comment by orbeckst** [comment]
Treat transformations as something that you get to add to a universe a single time. 

There is no API at the moment for modifying them in any form. 

Part of the reason is that a transformation applied to an in-memory Universe changes the underlying data. There is no safe way to undo the transformation and the apply the new one (or none) except for starting over fresh. In order to avoid the danger of working with corrupted data we don’t support changing transformations. 

In order to change this situation we would need to think about safe guards that prevent potential data corruption.

**Reply by BradyAJohnston** [reply]
Thanks for all the info! So the best approach to change the transformations applied would be to re-instantiate the Universe with the same starting topology and trajectory and then apply a different transformation if desired.

**Reply by orbeckst** [reply]
yes

**Reply by orbeckst** [reply]
`Universe.copy()` keeps transformations in the copy so something like
```python
u1 = Universe(TOPOL, TRAJ)
u2 = Universe(u1.filename, u1.trajectory.filename)
```
may cover simple cases.

**Reply by BradyAJohnston** [reply]
ah that's great, I didn't realise you could just grab the filenames from the universe. I think this should mostly cover what I have in mind!

**Reply by orbeckst** [reply]
yeah, just be aware that we sometimes cheat and `filename` is not a string, e.g., when we loaded a topology from a buffer/[NamedStream](https://docs.mdanalysis.org/stable/documentation_pages/lib/util.html#MDAnalysis.lib.util.NamedStream) (IIRC...)

**Reply by BradyAJohnston** [reply]
I'm thinking that at least for use in Molecular Nodes, that I might keep around the original untransformed universe, then create copies whenever I want to add / alter a transformation: 

```python
u = Universe(TOPOL, TRAJ)

u_smoothed = u.copy()
u_smoothed.trajectory.add_transformation()
```

I can see this maybe being problematic with particularly large universes though. Will have to do some testing.

**Reply by orbeckst** [reply]
EDIT: Keeping an unmodified copy sounds like a good idea. I think very few universes are so big that you can't do that.

EDIT: My original suggestion https://github.com/MDAnalysis/mdanalysis/discussions/4477#discussioncomment-8687162 to just use filenames may not be such a great ideas and it may better to do the following:

The safer approach is probably to reverse engineer `Universe.copy()` and the underlying `topology.copy()` and `trajectory.copy()` methods — the latter actively patches transformations back in so perhaps writing one's own `copy_universe_without_transformations(u)` method may be better in the long run (or raise an issue to add a `no_transformations` kwarg to `Universe.copy()`...)

**Reply by BradyAJohnston** [reply]
Ah OK I'll have a look at the copy method as well then. Thanks for all of the details!
